File: bindings/python/wlite/_statement.py
"""Python Statement wrapper."""

from wlite._ffi import lib, check_result, WliteError, WLITE_OK, WLITE_NOT_FOUND

class Statement:
    def __init__(self, ptr): self._ptr = ptr
    def bind_int64(self, i, v): check_result(lib.wlite_bind_int64(self._ptr, i, v))
    def bind_double(self, i, v): check_result(lib.wlite_bind_double(self._ptr, i, v))
    def bind_text(self, i, v): check_result(lib.wlite_bind_text(self._ptr, i, v.encode("utf-8")))
    def bind_null(self, i): check_result(lib.wlite_bind_null(self._ptr, i))
    def step(self):
        rc = lib.wlite_step(self._ptr)
        if rc == WLITE_OK: return True
        if rc == WLITE_NOT_FOUND: return False
        raise WliteError(rc)
    def column_count(self): return lib.wlite_column_count(self._ptr)
    def column_name(self, i): return lib.wlite_column_name(self._ptr, i).decode()
    def column_type(self, i): return lib.wlite_column_type(self._ptr, i)
    def column_int64(self, i): return lib.wlite_column_int64(self._ptr, i)
    def column_double(self, i): return lib.wlite_column_double(self._ptr, i)
    def column_text(self, i):
        p = lib.wlite_column_text(self._ptr, i)
        return p.decode() if p else ""
    def column_value(self, i):
        t = self.column_type(i)
        if t == 0: return None
        elif t == 1: return self.column_int64(i)
        elif t == 2: return self.column_double(i)
        elif t == 3: return self.column_text(i)
        return None
    def __iter__(self): return self
    def __next__(self):
        if self.step(): return [self.column_value(i) for i in range(self.column_count())]
        raise StopIteration
    def __del__(self):
        if self._ptr: lib.wlite_stmt_finalize(self._ptr)

File: bindings/rust/src/statement.rs
use std::ffi::CStr;

use crate::error::{check, WliteError};
use crate::ffi;
use crate::Result;

pub struct Statement {
    ptr: *mut ffi::WliteStmt,
}

unsafe impl Send for Statement {}

impl Statement {
    pub(crate) fn from_raw(ptr: *mut ffi::WliteStmt) -> Self {
        Self { ptr }
    }

    pub fn bind_int64(&self, index: i32, value: i64) -> Result<()> {
        check(unsafe { ffi::wlite_bind_int64(self.ptr, index, value) })
    }

    pub fn bind_double(&self, index: i32, value: f64) -> Result<()> {
        check(unsafe { ffi::wlite_bind_double(self.ptr, index, value) })
    }

    pub fn bind_text(&self, index: i32, value: &str) -> Result<()> {
        let c_value = std::ffi::CString::new(value).unwrap();
        check(unsafe { ffi::wlite_bind_text(self.ptr, index, c_value.as_ptr()) })
    }

    pub fn bind_null(&self, index: i32) -> Result<()> {
        check(unsafe { ffi::wlite_bind_null(self.ptr, index) })
    }

    pub fn step(&self) -> Result<bool> {
        let result = unsafe { ffi::wlite_step(self.ptr) };
        if result == ffi::WLITE_OK {
            Ok(true) // row available
        } else if result == ffi::WLITE_NOT_FOUND {
            Ok(false) // done
        } else {
            Err(WliteError::from_result(result))
        }
    }

    pub fn reset(&self) -> Result<()> {
        check(unsafe { ffi::wlite_step(self.ptr) })
    }

    pub fn column_count(&self) -> i32 {
        unsafe { ffi::wlite_column_count(self.ptr) }
    }

    pub fn column_name(&self, index: i32) -> &str {
        unsafe {
            let ptr = ffi::wlite_column_name(self.ptr, index);
            CStr::from_ptr(ptr).to_str().unwrap_or("")
        }
    }

    pub fn column_int64(&self, index: i32) -> i64 {
        unsafe { ffi::wlite_column_int64(self.ptr, index) }
    }

    pub fn column_double(&self, index: i32) -> f64 {
        unsafe { ffi::wlite_column_double(self.ptr, index) }
    }

    pub fn column_text(&self, index: i32) -> &str {
        unsafe {
            let ptr = ffi::wlite_column_text(self.ptr, index);
            if ptr.is_null() { "" } else { CStr::from_ptr(ptr).to_str().unwrap_or("") }
        }
    }
}

impl Drop for Statement {
    fn drop(&mut self) {
        unsafe { ffi::wlite_stmt_finalize(self.ptr) };
    }
}
